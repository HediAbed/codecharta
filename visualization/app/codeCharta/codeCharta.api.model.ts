import {
	AttributeTypes,
	AttributeTypeValue,
	CodeMapNode,
	Edge,
	MarkedPackage,
	MetricStatisticsItem,
	RecursivePartial,
	Settings
} from "./codeCharta.model"

export interface ExportCCFile {
	projectName: string
	apiVersion: string
	fileChecksum: string
	nodes: CodeMapNode[]
	attributeTypes?: AttributeTypes | OldAttributeTypes
	edges?: Edge[]
	markedPackages?: MarkedPackage[]
	blacklist?: ExportBlacklistItem[]
	metricStatistics?: ExportMetricStatisticsItem[]
}

export interface ExportBlacklistItem {
	path: string
	type: ExportBlacklistType
}

export type ExportMetricStatisticsItem = MetricStatisticsItem

export enum ExportBlacklistType {
	hide = "hide",
	exclude = "exclude"
}

export enum APIVersions {
	ZERO_POINT_ONE = "0.1",
	ONE_POINT_ZERO = "1.0",
	ONE_POINT_ONE = "1.1",
	ONE_POINT_TWO = "1.2"
}

export interface ExportScenario {
	name: string
	settings: RecursivePartial<Settings>
}

export interface OldAttributeTypes {
	nodes?: [{ [key: string]: AttributeTypeValue }?]
	edges?: [{ [key: string]: AttributeTypeValue }?]
}
