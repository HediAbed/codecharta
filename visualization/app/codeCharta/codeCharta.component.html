<tool-bar-component></tool-bar-component>

<file-extension-bar-component ></file-extension-bar-component>
<div style="margin-top:1px; width: 100%; min-height: 100px; position: fixed; z-index: 10; overflow: hidden; white-space: nowrap;font-size: 0px;" >	
	<ribbon-bar-component class="ribbon-bar-component"></ribbon-bar-component >
</div>


<div id="actionButtons">
	<md-button
		ng-if="$ctrl._viewModel.focusedNodePath.length > 0"
		class="md-fab md-primary dark_shadow_opacity long-button"
		aria-label="Remove Focus on node"
		ng-click="$ctrl.removeFocusedNode()"
	>
		Show complete map
	</md-button>
	<md-button class="md-fab md-primary dark_shadow_opacity" aria-label="Fit Map to View" ng-click="$ctrl.fitMapToView()">
		<i class="fa fa-compress fa-lg"></i>
	</md-button>
</div>

<code-map-component></code-map-component>

<legend-panel-component></legend-panel-component>

<detail-panel-component></detail-panel-component>

<loading-gif-file-component></loading-gif-file-component>

<div id="mw-logo">
	<div class="logo">
		<a href="https://www.maibornwolff.de" target="_blank"><img src="assets/logo.png" alt="MaibornWolff"/></a>
		<h2>
			<span>CodeCharta {{ $ctrl._viewModel.version }}</span>
		</h2>
	</div>
</div>

<node-context-menu-component></node-context-menu-component>
