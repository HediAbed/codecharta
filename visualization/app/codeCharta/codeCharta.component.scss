@import "../mixins.scss";

code-charta-component {
	$animationDuration: 0.5;

	#mw-logo {
		.logo {
			width: 160px;
			height: 100px;
			position: absolute;
			left: 0;
			bottom: 30px;
			z-index: 29;
		}

		a,
		img {
			width: 5em;
			display: block;
			margin: auto !important;
		}

		h2 {
			text-transform: uppercase;
			color: #5a585a;
			font-family: Roboto, sans-serif;
			font-weight: 500;
			font-size: 9pt;
			text-align: center;
			margin: 0;
			z-index: -1;
		}
	}

	#title {
		position: fixed;
		z-index: 2;
		width: 100%;
	}

	.small-info {
		color: rgba(0, 0, 0, 0.38);
		font-size: 9pt;
	}

	#codeMap {
		position: fixed;
		z-index: 0;
		top: 90px;
		left: 0;
		width: 100%;
		height: 100%;
		transition: all $animationDuration + s ease;

		&.expanded {
			height: calc(100% - 170px);
		}
	}

	#actionButtons {
		position: fixed;
		bottom: 2em;
		right: 2em;
		z-index: 60;

		i.fa-folder-open {
			margin: 4px 0 0 2px;
		}

		.long-button {
			width: inherit;
			padding-left: 10px;
			padding-right: 10px;
			border-radius: 100px;
		}

		> * {
			margin-left: 4px;
			margin-right: 4px;
			z-index: 60;
		}
	}

	.dark_shadow_opacity {
		box-shadow: 0 1px 5px 0 rgba(0, 0, 0, 0.2) !important;
		opacity: 0.97;
	}

	.bottom_shadow {
		box-shadow: 0 4px 4px -2px rgba(0, 0, 0, 0.2) !important;
	}

	.md-button.round-button {
		margin-right: 0 !important;
		float: right;
		min-width: 36px;
		width: 36px;
		border-radius: 50%;
	}

	.ribbon-bar-component {
		z-index: 10;

		md-card,
		.section .section-header,
		.section .section-body,
		#toggle-ribbon-bar-fab {
			transition: all $animationDuration + s ease;
		}
	}

	.ribbon-bar-wrapper {
		margin-top: 1px;
		width: 100%;
		min-height: 100px;
		height: 500px;
		position: fixed;
		z-index: 10;
		overflow: hidden;
		white-space: nowrap;
		font-size: 0px;
		pointer-events: none;
	}

	.angular-green {
		color: #00bfa5;
	}

	.small-action-button {
		height: 24px;
		width: 24px;
		float: left;

		*,
		.md-fab:active {
			margin: 0;
			padding: 0;
		}

		.md-fab {
			position: relative;
			min-height: 0;
			width: 100%;
			height: 100%;
			z-index: 0;
			border-radius: 50%;

			i {
				position: absolute;
				top: 5px;
				left: 6px;
				color: white;

				&:hover {
					cursor: pointer;
					color: #eee;
				}
			}
		}
	}
}
