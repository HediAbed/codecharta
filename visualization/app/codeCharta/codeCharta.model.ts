export interface CCFile {
  map: CodeMapNode;
  settings: {
      fileSettings: RecursivePartial<Settings>
  };
  fileMeta: FileMeta;
}

export interface CodeMapNode {
    name: string;
    type: string;
    children?: CodeMapNode[];
    attributes: {
        [key: string]: number
    };
    deltas?: {
        [key: string]: number
    };
    link?: string;
    origin?: string;
    visible?: boolean;
    path?: string;
    markingColor?: string;
}

export interface FileMeta {
  fileName: string;
  apiVersion: string;
  projectName: string;
}

export interface Settings {
    fileSettings: FileSettings;
    dynamicSettings: DynamicSettings;
    appSettings: AppSettings;
}

export interface FileSettings {
    attributeTypes: AttributeTypes;
    blacklist: Array<BlacklistItem>;
    edges: Edge[];
    markedPackages: MarkedPackages[];
}

export interface DynamicSettings {
    renderMode: RenderMode; // TODO remove from settings ?
    areaMetric: string;
    heightMetric: string;
    colorMetric: string;
    focusedNodePath: string;
    searchedNodePaths: Array<string>;
    searchPattern: string;
    margin: number;
    neutralColorRange: ColorRange;
}

export interface AppSettings {
    amountOfTopLabels: number;
    scaling: Vector3d;
    camera: Vector3d;
    deltaColorFlipped: boolean;
    enableEdgeArrows: boolean;
    hideFlatBuildings: boolean;
    maximizeDetailPanel: boolean;
    invertHeight: boolean;
    dynamicMargin: boolean;
    isWhiteBackground: boolean;
    mapColors: MapColors;
    whiteColorBuildings: boolean;
}

export interface MapColors {
    positive: string;
    neutral: string;
    negative: string;
    selected: string;
    defaultC: string;
    positiveDelta: string;
    negativeDelta: string;
    base: string;
    flat: string;
    lightGrey: string;
    angularGreen: string;
    markingColors: string[];
}

export interface ColorRange {
    from: number;
    to: number;
    flipped: boolean;
}

export interface Vector3d {
    x: number;
    y: number;
    z: number;
}

export interface AttributeTypes {
    nodes?: {
        [key: string]: AttributeType
    };
    edges?: {
        [key: string]: AttributeType
    };
}

export enum AttributeType {
    absolute, relative
}


export interface Edge {
    fromNodeName: string;
    toNodeName: string;
    attributes: {
        [key: string]: number
    };
    visible?: boolean;
}

export interface BlacklistItem {
    path: string;
    type: BlacklistType;
}

export enum BlacklistType {
    hide = "hide",
    exclude = "exclude"
}

export interface MarkedPackages {
    path: string,
    color: string,
    attributes: {
        [key: string]: any
    }
}

export enum RenderMode {
  Single = "Single",
  Multiple = "Multiple",
  Delta = "Delta",
}

export interface MetricData {
    name: string;
    maxValue: number;
}

export interface Scenario {
    name: string;
    settings: RecursivePartial<Settings>;
}

export interface UrlData {
    filenames: string[];
    settings: Partial<Settings>;
}

export type RecursivePartial<T> = {
    [P in keyof T]?: RecursivePartial<T[P]>;
};
