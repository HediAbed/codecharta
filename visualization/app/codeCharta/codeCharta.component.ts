import { NameDataPair, UrlUtils } from "./util/urlUtils"
import {IHttpService, ILocationService, IRootScopeService} from "angular"
import "./codeCharta.component.scss"
import { NodeContextMenuController } from "./ui/nodeContextMenu/nodeContextMenu.component"
import { CodeMapActionsService } from "./ui/codeMap/codeMap.actions.service"
import { CodeChartaService } from "./codeCharta.service"

/**
 * This is the main controller of the CodeCharta application
 */
export class CodeChartaController {
	public viewModel = {
		version: require("../../package.json").version,
		numberOfLoadingTasks: 0
	}
	private urlUtils: UrlUtils

	/* @ngInject */
	constructor(
		//private threeOrbitControlsService: ThreeOrbitControlsService,
		private $rootScope: IRootScopeService,
		//private dialogService: DialogService,
		//private codeMapActionsService: CodeMapActionsService,
		private codeChartaService: CodeChartaService,
		private $location: ILocationService,
		private $http: IHttpService
	) {
		this.urlUtils = new UrlUtils(this.$location, this.$http)
		this.subscribeToLoadingEvents(this.$rootScope)
		this.loadFileOrSample()
	}

	public fitMapToView() {
		// this.threeOrbitControlsService.autoFitTo()
	}

	public removeFocusedNode() {
		//this.codeMapActionsService.removeFocusedNode()
	}

	public loadFileOrSample() {
		this.viewModel.numberOfLoadingTasks++
		return this.urlUtils.getFileDataFromQueryParam().then((data: NameDataPair[]) => {
			if (data.length > 0) {
				this.tryLoadingFiles(data)
			} else {
				this.tryLoadingSampleFiles()
			}
		}, this.tryLoadingSampleFiles.bind(this))
	}

	public tryLoadingSampleFiles() {
		if (this.urlUtils.getParam("file")) {
			/*this.dialogService.showErrorDialog(
				"One or more files from the given file URL parameter could not be loaded. Loading sample files instead."
			)*/
		}
		this.tryLoadingFiles([
            { name: "sample1.json", data: require("./assets/sample1.json") },
            { name: "sample2.json", data: require("./assets/sample2.json") }
        ]);
    }
    
    private tryLoadingFiles(values: NameDataPair[]) {
        this.codeChartaService
        .loadFiles(
            values
        )
        .then(() => this.viewModel.numberOfLoadingTasks--)
        .catch(e => {
			this.viewModel.numberOfLoadingTasks--
			console.error(e);
            this.printErrors(e)
        })
    }

	public printErrors(errors: Object) {
		//this.dialogService.showErrorDialog(JSON.stringify(errors, null, "\t"))
	}

	private subscribeToLoadingEvents($rootScope: angular.IRootScopeService) {
		$rootScope.$on("add-loading-task", () => {
			this.viewModel.numberOfLoadingTasks++
		})

		$rootScope.$on("remove-loading-task", () => {
			this.viewModel.numberOfLoadingTasks--
		})
	}
}

export const codeChartaComponent = {
	selector: "codeChartaComponent",
	template: require("./codeCharta.component.html"),
	controller: CodeChartaController
}
