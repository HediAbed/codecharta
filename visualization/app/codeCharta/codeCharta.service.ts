import { FileValidator } from "./util/fileValidator"
import { CCFile } from "./codeCharta.model"
import { NameDataPair } from "./util/urlUtils"
import { SettingsService } from "./state/settings.service"
import { IRootScopeService } from "angular"
import { FileStateService } from "./state/fileState.service";
import {ThreeOrbitControlsService} from "./ui/codeMap/threeViewer/threeOrbitControlsService";

export class CodeChartaService {

	// TODO: use ROOT_NAME and ROOT_PATH everywhere in project instead of individual strings
	public static ROOT_NAME = "root"
	public static ROOT_PATH =  "/" + CodeChartaService.ROOT_NAME

	constructor(
		private $rootScope: IRootScopeService,
		private settingsService: SettingsService,
		private fileStateService: FileStateService,
		private threeOrbitControlsService: ThreeOrbitControlsService
	) {
	}

	public resetMaps(): any {
		throw new Error("Method not implemented.")
	}

	public loadFiles(nameDataPairs: NameDataPair[]): Promise<void> {
		return new Promise((resolve, reject) => {

			this.settingsService.updateSettings(this.settingsService.getDefaultSettings())

			nameDataPairs.forEach((nameDataPair: NameDataPair) => {
				const errors = FileValidator.validate(nameDataPair.content as any)
				if (errors.length === 0) {
					const ccFile = this.getCCFile(nameDataPair.fileName, nameDataPair.content)
					this.fileStateService.addFile(ccFile)
				} else {
					reject(errors)
				}
			})

			this.fileStateService.setSingle(this.fileStateService.getCCFiles()[0])
			// TODO this.settingsService.updateSettingsFromUrl();
			resolve()
		})
	}


	private getCCFile(fileName: string, fileContent: any): CCFile {
		return {
			fileMeta: {
				fileName: fileName,
				projectName: fileContent.projectName,
				apiVersion: fileContent.apiVersion
			},
			settings: {
				fileSettings: {
					edges: fileContent.edges || [],
					attributeTypes: fileContent.attributeTypes || {},
					blacklist: fileContent.blacklist || [],
				}
			},
			map: fileContent.nodes[0]
		}
	}
}
