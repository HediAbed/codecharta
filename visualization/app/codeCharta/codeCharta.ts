"use strict";

import "./core/core.module";
import "./ui/ui";

import {codeChartaComponent} from "./codeCharta.component";

import angular from "angular";
import { CodeChartaService } from "./codeCharta.service";
import _ from "lodash";

angular.module(
    "app.codeCharta",
    ["app.codeCharta.core", "app.codeCharta.ui"]
);

angular.module("app.codeCharta").component(
    codeChartaComponent.selector,
    codeChartaComponent
).service(
    _.camelCase(CodeChartaService.name),
    CodeChartaService
);


