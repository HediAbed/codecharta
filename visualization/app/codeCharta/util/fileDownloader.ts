import angular from "angular"
import * as d3 from "d3"
import { CodeMapNode, BlacklistType, BlacklistItem, FileSettings, ExportCCFile, FileMeta, AttributeTypes } from "../codeCharta.model"
import { DownloadCheckboxNames } from "../ui/dialog/dialog.download.component"
import { CodeChartaService } from "../codeCharta.service"

export class FileDownloader {
	public static downloadCurrentMap(
		map: CodeMapNode,
		fileMeta: FileMeta,
		fileSettings: FileSettings,
		downloadSettingsNames: string[],
		fileName: string
	) {
		const exportCCFile: ExportCCFile = this.getProjectDataAsCCJsonFormat(map, fileMeta, fileSettings, downloadSettingsNames)
		const newFileNameWithExtension: string = fileName + CodeChartaService.CC_FILE_EXTENSION
		this.downloadData(exportCCFile, newFileNameWithExtension)
	}

	private static getProjectDataAsCCJsonFormat(
		map: CodeMapNode,
		fileMeta: FileMeta,
		fileSettings: FileSettings,
		downloadSettingsNames: string[]
	): ExportCCFile {
		return {
			projectName: fileMeta.projectName,
			apiVersion: fileMeta.apiVersion,
			nodes: [this.removeJsonHashkeysAndVisibleAttribute(map)],
			attributeTypes: this.getAttributeTypesForJSON(fileSettings.attributeTypes),
			edges: downloadSettingsNames.includes(DownloadCheckboxNames.edges) ? fileSettings.edges : [],
			markedPackages: downloadSettingsNames.includes(DownloadCheckboxNames.markedPackages) ? fileSettings.markedPackages : [],
			blacklist: this.getBlacklistToDownload(downloadSettingsNames, fileSettings.blacklist)
		}
	}

	private static getBlacklistToDownload(downloadSettingsNames: string[], blacklist: BlacklistItem[]): BlacklistItem[] {
		let mergedBlacklist: BlacklistItem[] = []

		if (downloadSettingsNames.includes(DownloadCheckboxNames.hides)) {
			mergedBlacklist.push(...this.getFilteredBlacklist(blacklist, BlacklistType.hide))
		}

		if (downloadSettingsNames.includes(DownloadCheckboxNames.excludes)) {
			mergedBlacklist.push(...this.getFilteredBlacklist(blacklist, BlacklistType.exclude))
		}
		return mergedBlacklist
	}

	private static getAttributeTypesForJSON(attributeTypes: AttributeTypes): AttributeTypes | {} {
		if (attributeTypes.edges.length === 0 && attributeTypes.nodes.length === 0) {
			return {}
		} else {
			return attributeTypes
		}
	}

	private static getFilteredBlacklist(blacklist: BlacklistItem[], type: BlacklistType): BlacklistItem[] {
		return blacklist.filter(x => x.type == type)
	}

	private static removeJsonHashkeysAndVisibleAttribute(map: CodeMapNode) {
		let copy: CodeMapNode = JSON.parse(JSON.stringify(map))
		d3.hierarchy(copy).each(node => {
			delete node.data.visible
		})
		return copy
	}

	private static downloadData(data, fileName: string) {
		let dataJson = data
		if (typeof data === "object") {
			dataJson = angular.toJson(data, 4)
		}

		const blob = new Blob([dataJson], { type: "text/json" })
		const e = document.createEvent("MouseEvents")
		const a = document.createElement("a")

		a.download = fileName
		a.href = window.URL.createObjectURL(blob)
		a.dataset.downloadurl = ["text/json", a.download, a.href].join(":")
		e.initMouseEvent("click", true, false, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null)
		a.dispatchEvent(e)
	}
}
