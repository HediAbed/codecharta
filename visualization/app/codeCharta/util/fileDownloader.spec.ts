import {stubDate} from "../../../mocks/dateMock.helper";
import {FileDownloader} from "./fileDownloader";
import {CCFile} from "../codeCharta.model";

describe("app.codeCharta.core.fileDownloader", function() {

    const file: CCFile = {
        fileMeta: {
            projectName: "myProjectName",
            fileName: "myFilename",
            apiVersion: "1.1.1",
        },
        map: {
            name: "root",
            type: "Folder",
            path: "/root",
            visible: true,
            attributes: {},
            children: [
                {
                    name: "a",
                    type: "Folder",
                    path: "/root/a",
                    visible: false,
                    attributes: {},
                    children: [
                        {
                            name: "aa",
                            type: "File",
                            path: "/root/a/aa",
                            visible: true,
                            attributes: {},
                        }
                    ]
                },
                {
                    name: "b",
                    type: "File",
                    path: "/root/b",
                    visible: false,
                    attributes: {},
                }
            ]
        },
        settings: {
            fileSettings: {
                edges: "myEdges",
                attributeTypes: "myAttributeTypes",
            }
        }
    }

    stubDate(new Date('2018-12-14T09:39:59'));

    it("should add json extension to given string", function() {
        expect(FileDownloader.addJsonFileEndingIfNecessary("fileSettings")).toBe("fileSettings.json");
    });

    it("should keep json extension with given string", function() {
        expect(FileDownloader.addJsonFileEndingIfNecessary("fileSettings.json")).toBe("fileSettings.json");
    });

    it("should add date string before json extension", function() {
        const filenameWithDate = FileDownloader.addDateToFileName("fileSettings.json");
        const regexp = new RegExp('^fileSettings\.\\d{2}_\\d{2}_\\d{4}\.json$');
        expect(regexp.test(filenameWithDate)).toBeTruthy();
    });

    it("should remove visible attribute from settings.map", function() {
        expect(FileDownloader.removeJsonHashkeysAndVisibleAttribute(file.map)).toMatchSnapshot();
    });

    it("should return correct project json content", function() {
        FileDownloader.downloadData = jest.fn(function() {return true;});
        expect(FileDownloader.getProjectDataAsCCJsonFormat(file)).toMatchSnapshot();
    });

});