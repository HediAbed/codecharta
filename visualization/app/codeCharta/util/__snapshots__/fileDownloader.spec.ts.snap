// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app.codeCharta.core.fileDownloader should remove visible attribute from settings.map 1`] = `
Object {
  "attributes": Object {},
  "children": Array [
    Object {
      "attributes": Object {},
      "children": Array [
        Object {
          "attributes": Object {},
          "name": "aa",
          "path": "/root/a/aa",
          "type": "File",
        },
      ],
      "name": "a",
      "path": "/root/a",
      "type": "Folder",
    },
    Object {
      "attributes": Object {},
      "name": "b",
      "path": "/root/b",
      "type": "File",
    },
  ],
  "name": "root",
  "path": "/root",
  "type": "Folder",
}
`;
